File: src/harness_core/sandbox.py
"""Sandbox Manager for Stage 4 — file-level isolation with claim tracking."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Sandbox:
    sandbox_id: str
    task_id: str
    status: str  # "created" | "active" | "released" | "failed"
    claimed_files: tuple[str, ...]
    created_at: str
    released_at: str | None = None


@dataclass(frozen=True)
class FileClaim:
    claim_id: str
    sandbox_id: str
    file_path: str
    claimed_at: str
    released: bool = False


@dataclass(frozen=True)
class ConflictReport:
    has_conflict: bool
    conflicting_sandbox_id: str | None
    conflicting_file: str | None
    message: str


class SandboxManager:
    """Track sandbox file claims and detect conflicts."""

    def __init__(self) -> None:
        self._sandboxes: dict[str, Sandbox] = {}
        self._file_owners: dict[str, str] = {}  # file_path -> sandbox_id
        self._claims: dict[str, FileClaim] = {}  # claim_id -> FileClaim
        self._claim_counter: int = 0

    def create_sandbox(
        self, task_id: str, files: tuple[str, ...], timestamp: str = "2026-01-01T00:00:00Z"
    ) -> Sandbox:
        sandbox_id = f"sbx_{task_id}_{len(self._sandboxes)}"
        sandbox = Sandbox(
            sandbox_id=sandbox_id,
            task_id=task_id,
            status="created",
            claimed_files=(),
            created_at=timestamp,
        )
        self._sandboxes[sandbox_id] = sandbox
        if files:
            report = self.claim_files(sandbox_id, files, timestamp)
            if report.has_conflict:
                raise ValueError(f"initial claim failed: {report.message}")
        return self._sandboxes[sandbox_id]

    def claim_files(
        self,
        sandbox_id: str,
        files: tuple[str, ...],
        timestamp: str = "2026-01-01T00:00:00Z",
    ) -> ConflictReport:
        sandbox = self._sandboxes.get(sandbox_id)
        if sandbox is None:
            return ConflictReport(True, None, None, f"unknown sandbox: {sandbox_id}")
        if sandbox.status not in ("created", "active"):
            return ConflictReport(
                True, None, None, f"sandbox {sandbox_id} is {sandbox.status}"
            )
        for file_path in files:
            owner = self._file_owners.get(file_path)
            if owner is not None and owner != sandbox_id:
                return ConflictReport(
                    True, owner, file_path, f"file {file_path} claimed by {owner}"
                )
        for file_path in files:
            if file_path not in self._file_owners:
                self._file_owners[file_path] = sandbox_id
                self._claim_counter += 1
                claim_id = f"claim_{self._claim_counter}"
                self._claims[claim_id] = FileClaim(
                    claim_id=claim_id,
                    sandbox_id=sandbox_id,
                    file_path=file_path,
                    claimed_at=timestamp,
                )
        existing = set(sandbox.claimed_files)
        existing.update(files)
        self._sandboxes[sandbox_id] = Sandbox(
            sandbox_id=sandbox.sandbox_id,
            task_id=sandbox.task_id,
            status="active",
            claimed_files=tuple(sorted(existing)),
            created_at=sandbox.created_at,
            released_at=sandbox.released_at,
        )
        return ConflictReport(False, None, None, "ok")

    def release_sandbox(
        self, sandbox_id: str, timestamp: str = "2026-01-01T00:00:00Z"
    ) -> Sandbox:
        sandbox = self._sandboxes.get(sandbox_id)
        if sandbox is None:
            raise ValueError(f"unknown sandbox: {sandbox_id}")
        for file_path in sandbox.claimed_files:
            if self._file_owners.get(file_path) == sandbox_id:
                del self._file_owners[file_path]
        for claim in self._claims.values():
            if claim.sandbox_id == sandbox_id and not claim.released:
                self._claims[claim.claim_id] = FileClaim(
                    claim_id=claim.claim_id,
                    sandbox_id=claim.sandbox_id,
                    file_path=claim.file_path,
                    claimed_at=claim.claimed_at,
                    released=True,
                )
        released = Sandbox(
            sandbox_id=sandbox.sandbox_id,
            task_id=sandbox.task_id,
            status="released",
            claimed_files=sandbox.claimed_files,
            created_at=sandbox.created_at,
            released_at=timestamp,
        )
        self._sandboxes[sandbox_id] = released
        return released

    def get_sandbox(self, sandbox_id: str) -> Sandbox | None:
        return self._sandboxes.get(sandbox_id)

    def list_active(self) -> tuple[Sandbox, ...]:
        return tuple(
            s for s in self._sandboxes.values() if s.status in ("created", "active")
        )

    def list_all(self) -> tuple[Sandbox, ...]:
        return tuple(self._sandboxes.values())

    def is_file_claimed(self, file_path: str) -> str | None:
        return self._file_owners.get(file_path)

    def get_claims(self, sandbox_id: str) -> tuple[FileClaim, ...]:
        return tuple(
            c
            for c in self._claims.values()
            if c.sandbox_id == sandbox_id
        )

File: engine/src/harness/sandbox.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Sandbox {
    pub sandbox_id: String,
    pub task_id: String,
    pub status: String,
    pub claimed_files: Vec<String>,
    pub created_at: String,
    pub released_at: Option<String>,
}

impl Default for Sandbox {
    fn default() -> Self {
        Self {
            sandbox_id: String::new(),
            task_id: String::new(),
            status: "created".to_string(),
            claimed_files: Vec::new(),
            created_at: "2026-01-01T00:00:00Z".to_string(),
            released_at: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FileClaim {
    pub claim_id: String,
    pub sandbox_id: String,
    pub file_path: String,
    pub claimed_at: String,
    pub released: bool,
}

impl Default for FileClaim {
    fn default() -> Self {
        Self {
            claim_id: String::new(),
            sandbox_id: String::new(),
            file_path: String::new(),
            claimed_at: "2026-01-01T00:00:00Z".to_string(),
            released: false,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ConflictReport {
    pub has_conflict: bool,
    pub conflicting_sandbox_id: Option<String>,
    pub conflicting_file: Option<String>,
    pub message: String,
}

impl Default for ConflictReport {
    fn default() -> Self {
        Self {
            has_conflict: false,
            conflicting_sandbox_id: None,
            conflicting_file: None,
            message: "ok".to_string(),
        }
    }
}

pub struct SandboxManager {
    sandboxes: HashMap<String, Sandbox>,
    file_owners: HashMap<String, String>,
    claims: HashMap<String, FileClaim>,
    claim_counter: u64,
}

impl Default for SandboxManager {
    fn default() -> Self {
        Self::new()
    }
}

impl SandboxManager {
    pub fn new() -> Self {
        Self {
            sandboxes: HashMap::new(),
            file_owners: HashMap::new(),
            claims: HashMap::new(),
            claim_counter: 0,
        }
    }

    pub fn create_sandbox(
        &mut self,
        task_id: &str,
        files: &[String],
        timestamp: &str,
    ) -> Result<Sandbox, String> {
        let sandbox_id = format!("sbx_{}_{}", task_id, self.sandboxes.len());
        let sandbox = Sandbox {
            sandbox_id: sandbox_id.clone(),
            task_id: task_id.to_string(),
            status: "created".to_string(),
            claimed_files: Vec::new(),
            created_at: timestamp.to_string(),
            released_at: None,
        };
        self.sandboxes.insert(sandbox_id.clone(), sandbox);
        if !files.is_empty() {
            let report = self.claim_files(&sandbox_id, files, timestamp)?;
            if report.has_conflict {
                return Err(format!("initial claim failed: {}", report.message));
            }
        }
        Ok(self.sandboxes.get(&sandbox_id).unwrap().clone())
    }

    pub fn claim_files(
        &mut self,
        sandbox_id: &str,
        files: &[String],
        timestamp: &str,
    ) -> Result<ConflictReport, String> {
        let sandbox = match self.sandboxes.get(sandbox_id) {
            Some(s) => s.clone(),
            None => {
                return Ok(ConflictReport {
                    has_conflict: true,
                    conflicting_sandbox_id: None,
                    conflicting_file: None,
                    message: format!("unknown sandbox: {}", sandbox_id),
                });
            }
        };
        if sandbox.status != "created" && sandbox.status != "active" {
            return Ok(ConflictReport {
                has_conflict: true,
                conflicting_sandbox_id: None,
                conflicting_file: None,
                message: format!("sandbox {} is {}", sandbox_id, sandbox.status),
            });
        }
        for file_path in files {
            if let Some(owner) = self.file_owners.get(file_path) {
                if owner != sandbox_id {
                    return Ok(ConflictReport {
                        has_conflict: true,
                        conflicting_sandbox_id: Some(owner.clone()),
                        conflicting_file: Some(file_path.clone()),
                        message: format!("file {} claimed by {}", file_path, owner),
                    });
                }
            }
        }
        for file_path in files {
            if !self.file_owners.contains_key(file_path) {
                self.file_owners
                    .insert(file_path.clone(), sandbox_id.to_string());
                self.claim_counter += 1;
                let claim_id = format!("claim_{}", self.claim_counter);
                self.claims.insert(
                    claim_id.clone(),
                    FileClaim {
                        claim_id,
                        sandbox_id: sandbox_id.to_string(),
                        file_path: file_path.clone(),
                        claimed_at: timestamp.to_string(),
                        released: false,
                    },
                );
            }
        }
        let mut existing: Vec<String> = sandbox.claimed_files.clone();
        for f in files {
            if !existing.contains(f) {
                existing.push(f.clone());
            }
        }
        existing.sort();
        let updated = Sandbox {
            sandbox_id: sandbox.sandbox_id.clone(),
            task_id: sandbox.task_id.clone(),
            status: "active".to_string(),
            claimed_files: existing,
            created_at: sandbox.created_at.clone(),
            released_at: sandbox.released_at.clone(),
        };
        self.sandboxes.insert(sandbox_id.to_string(), updated);
        Ok(ConflictReport {
            has_conflict: false,
            conflicting_sandbox_id: None,
            conflicting_file: None,
            message: "ok".to_string(),
        })
    }

    pub fn release_sandbox(
        &mut self,
        sandbox_id: &str,
        timestamp: &str,
    ) -> Result<Sandbox, String> {
        let sandbox = match self.sandboxes.get(sandbox_id) {
            Some(s) => s.clone(),
            None => return Err(format!("unknown sandbox: {}", sandbox_id)),
        };
        for file_path in &sandbox.claimed_files {
            if self.file_owners.get(file_path).map(|s| s.as_str()) == Some(sandbox_id) {
                self.file_owners.remove(file_path);
            }
        }
        let claim_ids: Vec<String> = self
            .claims
            .values()
            .filter(|c| c.sandbox_id == sandbox_id && !c.released)
            .map(|c| c.claim_id.clone())
            .collect();
        for cid in claim_ids {
            if let Some(claim) = self.claims.get(&cid) {
                self.claims.insert(
                    cid.clone(),
                    FileClaim {
                        claim_id: claim.claim_id.clone(),
                        sandbox_id: claim.sandbox_id.clone(),
                        file_path: claim.file_path.clone(),
                        claimed_at: claim.claimed_at.clone(),
                        released: true,
                    },
                );
            }
        }
        let released = Sandbox {
            sandbox_id: sandbox.sandbox_id.clone(),
            task_id: sandbox.task_id.clone(),
            status: "released".to_string(),
            claimed_files: sandbox.claimed_files.clone(),
            created_at: sandbox.created_at.clone(),
            released_at: Some(timestamp.to_string()),
        };
        self.sandboxes
            .insert(sandbox_id.to_string(), released.clone());
        Ok(released)
    }

    pub fn get_sandbox(&self, sandbox_id: &str) -> Option<&Sandbox> {
        self.sandboxes.get(sandbox_id)
    }

    pub fn list_active(&self) -> Vec<&Sandbox> {
        self.sandboxes
            .values()
            .filter(|s| s.status == "created" || s.status == "active")
            .collect()
    }

    pub fn list_all(&self) -> Vec<&Sandbox> {
        self.sandboxes.values().collect()
    }

    pub fn is_file_claimed(&self, file_path: &str) -> Option<&str> {
        self.file_owners.get(file_path).map(|s| s.as_str())
    }

    pub fn get_claims(&self, sandbox_id: &str) -> Vec<&FileClaim> {
        self.claims
            .values()
            .filter(|c| c.sandbox_id == sandbox_id)
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_sandbox_no_files() {
        let mut mgr = SandboxManager::new();
        let sb = mgr
            .create_sandbox("task1", &[], "2026-01-01T00:00:00Z")
            .unwrap();
        assert_eq!(sb.task_id, "task1");
        assert_eq!(sb.status, "created");
        assert!(sb.claimed_files.is_empty());
        assert!(sb.released_at.is_none());
    }

    #[test]
    fn test_create_sandbox_with_files() {
        let mut mgr = SandboxManager::new();
        let files = vec!["a.py".to_string(), "b.py".to_string()];
        let sb = mgr
            .create_sandbox("task1", &files, "2026-01-01T00:00:00Z")
            .unwrap();
        assert_eq!(sb.status, "active");
        assert_eq!(sb.claimed_files, vec!["a.py", "b.py"]);
    }

    #[test]
    fn test_file_conflict_on_create() {
        let mut mgr = SandboxManager::new();
        let files = vec!["shared.py".to_string()];
        mgr.create_sandbox("task1", &files, "2026-01-01T00:00:00Z")
            .unwrap();
        let result = mgr.create_sandbox("task2", &files, "2026-01-01T00:00:00Z");
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("initial claim failed"));
    }

    #[test]
    fn test_release_sandbox_frees_files() {
        let mut mgr = SandboxManager::new();
        let files = vec!["x.rs".to_string()];
        let sb = mgr
            .create_sandbox("t1", &files, "2026-01-01T00:00:00Z")
            .unwrap();
        let sbx_id = sb.sandbox_id.clone();
        let released = mgr
            .release_sandbox(&sbx_id, "2026-01-01T01:00:00Z")
            .unwrap();
        assert_eq!(released.status, "released");
        assert!(released.released_at.is_some());
        assert!(mgr.is_file_claimed("x.rs").is_none());
    }

    #[test]
    fn test_claim_files_on_active_sandbox() {
        let mut mgr = SandboxManager::new();
        let sb = mgr
            .create_sandbox("t1", &[], "2026-01-01T00:00:00Z")
            .unwrap();
        let sbx_id = sb.sandbox_id.clone();
        let report = mgr
            .claim_files(&sbx_id, &["new.py".to_string()], "2026-01-01T00:01:00Z")
            .unwrap();
        assert!(!report.has_conflict);
        assert_eq!(report.message, "ok");
        assert_eq!(mgr.is_file_claimed("new.py"), Some(sbx_id.as_str()));
    }

    #[test]
    fn test_claim_files_on_released_sandbox_returns_conflict() {
        let mut mgr = SandboxManager::new();
        let sb = mgr
            .create_sandbox("t1", &[], "2026-01-01T00:00:00Z")
            .unwrap();
        let sbx_id = sb.sandbox_id.clone();
        mgr.release_sandbox(&sbx_id, "2026-01-01T01:00:00Z")
            .unwrap();
        let report = mgr
            .claim_files(&sbx_id, &["f.py".to_string()], "2026-01-01T02:00:00Z")
            .unwrap();
        assert!(report.has_conflict);
        assert!(report.message.contains("released"));
    }

    #[test]
    fn test_list_active_and_list_all() {
        let mut mgr = SandboxManager::new();
        mgr.create_sandbox("t1", &[], "2026-01-01T00:00:00Z")
            .unwrap();
        let sb2 = mgr
            .create_sandbox("t2", &[], "2026-01-01T00:00:00Z")
            .unwrap();
        mgr.release_sandbox(&sb2.sandbox_id, "2026-01-01T01:00:00Z")
            .unwrap();
        assert_eq!(mgr.list_active().len(), 1);
        assert_eq!(mgr.list_all().len(), 2);
    }

    #[test]
    fn test_get_claims_for_sandbox() {
        let mut mgr = SandboxManager::new();
        let files = vec!["a.py".to_string(), "b.py".to_string()];
        let sb = mgr
            .create_sandbox("t1", &files, "2026-01-01T00:00:00Z")
            .unwrap();
        let claims = mgr.get_claims(&sb.sandbox_id);
        assert_eq!(claims.len(), 2);
        for c in &claims {
            assert!(!c.released);
        }
    }

    #[test]
    fn test_release_unknown_sandbox_errors() {
        let mut mgr = SandboxManager::new();
        let result = mgr.release_sandbox("nonexistent", "2026-01-01T00:00:00Z");
        assert!(result.is_err());
    }

    #[test]
    fn test_sandbox_default_values() {
        let sb = Sandbox::default();
        assert_eq!(sb.status, "created");
        assert!(sb.claimed_files.is_empty());
        assert!(sb.released_at.is_none());
    }
}
